### PR TITLE
googleSignIn API

### PR DESCRIPTION
**Pull request by NimishaGarg**

### CODE REVIEW

File: src/main/java/helper/GoogleAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        PeopleService peopleService =
+                new PeopleService.Builder(httpTransport, jsonFactory, requestInitializer).build();
+        Person profile = peopleService.people().get("people/" + userId)
+                .setPersonFields("phoneNumbers")
+                .execute();
+        List<PhoneNumber> phoneNumberList = profile.getPhoneNumbers();
+        if(phoneNumberList.isEmpty()) {
+            return null;
+        }
+        return phoneNumberList.get(0).getValue();
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleAuthenticator.java:72]: Just curious: Why are we just going with the first phone number? What about the others?
Consider leaving a comment about this here.

File: src/main/java/helper/GoogleUserMapper.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package helper;
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:1]: Instead of just having the Google Auth classes in the package `helper`, let's have them in their own package/sub-package?
WDYT @sankalpbansal-google ?

File: src/main/java/helper/GoogleUserMapper.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Base64;
+
+public class GoogleUserMapper {
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:10]: final?

File: src/main/java/helper/GoogleUserMapper.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    public String getUsernameFromEmail(String email) {
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:11]: This can be static.

File: src/main/java/helper/GoogleUserMapper.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        int usernameEndsAt = email.indexOf('@');
+        if(usernameEndsAt == -1) {
+            return null;
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:14]: Why return null here?
Shouldn't it be an error here for an email to not have the character `@` in it?
Perhaps we can throw `IllegalArgumentException`.

File: src/main/java/helper/GoogleUserMapper.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        return email.substring(0, usernameEndsAt);
+    }
+
+    public String getPictureFromUrl(String pictureUrl) throws IOException {
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:19]: static

File: src/main/java/helper/GoogleUserMapper.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        byte[] buffer = new byte[1024];
+        int read = 0;
+        while((read = inputStream.read(buffer, 0, buffer.length)) != -1) {
+            baos.write(buffer, 0, read);
+        }
+        baos.flush();
+        Base64.Encoder encoder = Base64.getEncoder();
+        return encoder.encodeToString(baos.toByteArray());
+    }
+}
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:33]: Please add a private constructor to this class. We don't need to instantiate it to use its static member functions.

File: src/main/java/helper/GoogleUserMapper.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        }
+        return email.substring(0, usernameEndsAt);
+    }
+
+    public String getPictureFromUrl(String pictureUrl) throws IOException {
+        URL imageUrl = new URL(pictureUrl);
+        URLConnection connection = imageUrl.openConnection();
+        InputStream inputStream = connection.getInputStream();
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        byte[] buffer = new byte[1024];
```
Comment (AgamAgarwal) [inline src/main/java/helper/GoogleUserMapper.java:24]: We are just reading the first 1024 bytes, i.e. 1 KB. I don't think that's enough for a picture. :)

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 9]
```diff
@@ -30,17 +30,19 @@ public long insert(User user) {
...
                                 .to(emailID)
                                 .build();
-        return !spannerTemplate
-                .query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true)) //setAllowPartialRead for reading specific columns
-                .isEmpty();
+        List<User> resultSet = spannerTemplate.query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true));
+        if(resultSet.isEmpty()) {
+            return null;
+        }
+        return resultSet.get(0);
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java:45]: This function is being used at 2 places-
1. One inside the native signup API, so for that case it won't matter which user is returned (because it just checks if user is not null the throw `UserAlreadyExisitsException`).
2. Second, in `getUser` function of `GoogleAuthenticator` class. Here we want that only if there exists a unique entry in User table with both the username and email matching then that user should be returned.
I will update this function accordingly.

[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java:45]: Actually I need to return the UserID if an existing user is found, will update accordingly.

File: src/main/java/helper/GoogleAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+public class GoogleAuthenticator {
+
+    @Autowired
+    private GoogleUserMapper googleUserMapper;
+
+    @Autowired
+    private UserAccessor userAccessor;
+
+    private static final String CLIENT_SECRET_FILE =
+        "src/main/resources/client_secret_731468299245-enaig2iil558lcjp15jmejvpphn5vpg5.apps.googleusercontent.com.json";
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java:34]: The token in the name is not the secret part, secret part is different and only present inside the json file. So this should be fine :)

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 11]
```diff
@@ -30,17 +30,19 @@ public long insert(User user) {
...
                                 .to(username)
                                 .bind("EmailID")
                                 .to(emailID)
                                 .build();
-        return !spannerTemplate
-                .query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true)) //setAllowPartialRead for reading specific columns
-                .isEmpty();
+        List<User> resultSet = spannerTemplate.query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true));
+        if(resultSet.isEmpty()) {
+            return null;
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done `Optional<Long>` return type for `getUserIdFromUsernameAndEmail`.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 12]
```diff
@@ -30,17 +30,19 @@ public long insert(User user) {
     }
 
     /* Checks if a user with the given username or email-id already exists in User table */
-    public boolean checkIfUserExists(String username, String emailID) {
+    public User checkIfUserExists(String username, String emailID) {
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 13]
```diff
@@ -30,17 +30,19 @@ public long insert(User user) {
     }
 
     /* Checks if a user with the given username or email-id already exists in User table */
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/helper/GoogleAuthenticator.java [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    private static final String REDIRECT_URI = "";
+    private final GoogleTokenResponse tokenResponse;
+
+    public GoogleAuthenticator(String authCode) throws IOException {
+        GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(
+                JacksonFactory.getDefaultInstance(), new FileReader(CLIENT_SECRET_FILE));
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java]: Done!

File: src/main/java/helper/GoogleAuthenticator.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+                    JacksonFactory.getDefaultInstance(),
+                    "https://oauth2.googleapis.com/token",
+                    clientSecrets.getDetails().getClientId(),
+                    clientSecrets.getDetails().getClientSecret(),
+                    authCode,
+                    REDIRECT_URI)
+                    .execute();
+    }
+
+    public GoogleCredentials getCredentials() throws IOException {
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java]: Yes, Done.

File: src/main/java/helper/GoogleAuthenticator.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        HttpTransport httpTransport = new NetHttpTransport();
+        JacksonFactory jsonFactory = new JacksonFactory();
+        PeopleService peopleService =
+                new PeopleService.Builder(httpTransport, jsonFactory, requestInitializer).build();
+        Person profile = peopleService.people().get("people/" + userId)
+                .setPersonFields("phoneNumbers")
+                .execute();
+        List<PhoneNumber> phoneNumberList = profile.getPhoneNumbers();
+        if(phoneNumberList.isEmpty()) {
+            return null;
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/helper/GoogleAuthenticator.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+import com.google.auth.oauth2.GoogleCredentials;
+import dbaccessor.user.UserAccessor;
+import entity.User;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.List;
+
+public class GoogleAuthenticator {
```
[the remark this answers is not in the document]
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java]: Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 3 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java]: Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 4 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java]: Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 5 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java]: Yes, Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 6 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java]: Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 7 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java:38]: Done.

File: src/main/java/helper/GoogleUserMapper.java (hunk 8 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java:30]: We are reading in blocks of 1KB, should I increase this block size? What do you suggest would be a good block size?

File: src/main/java/helper/GoogleAuthenticator.java (hunk 1 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleAuthenticator.java]: No particular reason. [Here](https://developers.google.com/people/api/rest/v1/people#phonenumber) are the details of the phoneNumber field, is there some way in which we can prefer one phone number over the other>

File: src/main/java/helper/GoogleUserMapper.java (hunk 2 above)
Comment (NimishaGarg) [reply src/main/java/helper/GoogleUserMapper.java]: Made a new package named `googlesignin`.

File: src/main/java/controller/UserAPI.java [hunk 18]
```diff
@@ -59,20 +64,20 @@
 
         String path = request.getRequestURI();
         requestValidator.signupRequestValidator(data, path);
-        /* necessary fields */
+        /** necessary fields */
```
Comment (AgamAgarwal) [inline src/main/java/controller/UserAPI.java:67]: This is not javadoc, but an implementation comment.
You should use `// ... ` or `/* ... */` style for implementation comments.

For small implementation comments, like this one, prefer using the single line variant - `// ...`

File: src/main/java/controller/UserAPI.java [hunk 19]
```diff
@@ -59,20 +64,20 @@
...
         String username = data.get("Username").toString();
         String emailID = data.get("EmailID").toString();
         String password = data.get("Password").toString();
         String mobileNo = data.get("MobileNo").toString();
 
-        /* optional field */
+        /** optional field */
```
Comment (AgamAgarwal) [inline src/main/java/controller/UserAPI.java:73]: same here

File: src/main/java/controller/UserAPI.java [hunk 20]
```diff
@@ -59,20 +64,20 @@
...
         String base64Image = "";
         if(data.get("Picture") != null){
             base64Image = data.get("Picture").toString();
         }
 
-        /* Checks if Username or email exists - throws exception if it does */
-        if (userAccessor.checkIfUserExists(username, emailID) != null) {
+        /** Checks if Username or email exists - throws exception if it does */
```
Comment (AgamAgarwal) [inline src/main/java/controller/UserAPI.java:79]: and here

File: src/main/java/googlesignin/GoogleUser.java [hunk 21]
```diff
@@ -67,27 +48,41 @@ public String getPhoneNumber(String userId) throws IOException {
...
         GoogleIdToken idToken = tokenResponse.parseIdToken();
         GoogleIdToken.Payload payload = idToken.getPayload();
         String email = payload.getEmail();
-        String username = googleUserMapper.getUsernameFromEmail(email);
-        User existingUser = userAccessor.checkIfUserExists(username, email);
-        if(existingUser != null) {
-            return existingUser;
+        String username = GoogleUserMapper.getUsernameFromEmail(email);
+        Optional<Long> existingUser = userAccessor.getUserIdFromUsernameAndEmail(username, email);
+        if(existingUser.isPresent()) {
```
Comment (AgamAgarwal) [inline src/main/java/googlesignin/GoogleUser.java:69]: What if a user exists with the same email but not same username, or same username but not same email?
In that case, `existingUser` will be empty as the function checks for `AND` and we will proceed to create a new user with either a conflicting email or a conflicting username.

Please fix this bug.

File: src/main/java/entity/User.java [hunk 22]
```diff
@@ -76,6 +76,10 @@ public User(
     this(-1, username, null, emailID, mobileNo, base64Image);
   }
 
+  public User(String username, String email, String picture) {
```
Comment (AgamAgarwal) [inline src/main/java/entity/User.java:79]: With so many constructors here, why not use static factory methods or better yet, Builder pattern. (This can be a separate PR too since this is independent of this change.)

File: src/main/java/googlesignin/GoogleUserMapper.java [hunk 23]
```diff
@@ -7,16 +7,22 @@
...
 import java.util.Base64;
 
-public class GoogleUserMapper {
-    public String getUsernameFromEmail(String email) {
+public final class GoogleUserMapper {
+    private static Base64.Encoder encoder;
+
+    private GoogleUserMapper() {
+        encoder = Base64.getEncoder();
+    }
```
Comment (AgamAgarwal) [inline src/main/java/googlesignin/GoogleUserMapper.java:15]: This is incorrect.
`encoder` is a static member but is initialised whenever an object of the class is created and not before that.
This is error prone.
If someone calls one of the static methods before an object is created, `encoder` will be set to null and a `NullPointerException` will be thrown.

You could instead do:
```
private static final Base64.Encoder encoder = Base64.getEncoder();

private GoogleUserMapper() {}
```

Or rather (preferred), leave `encoder` to be a local variable in the method `getPictureFromUrl`. `Base64.getEncoder` is a "fast" method in the sense that it does not create new objects for each call but returns the same object. So calling it in that method is safe and efficient, you need not store it in a static field here.

File: src/main/java/helper/GoogleUserMapper.java (hunk 8 above)
Comment (AgamAgarwal) [reply src/main/java/helper/GoogleUserMapper.java]: Oh sorry, I didn't realise this was block by block read. My bad.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 24]
```diff
@@ -21,29 +22,48 @@
...
         user.setUserId(id);
         spannerTemplate.insert(user);
         return id;
     }
 
-    /* Checks if a user with the given username or email-id already exists in User table */
-    public boolean checkIfUserExists(String username, String emailID) {
-        String SQLStatment = "SELECT Username FROM User WHERE Username=@Username OR EmailID=@EmailID";
+    /** Gets the UserID of the user having the given username and email-id */
+    public Optional<Long> getUserIdFromUsernameAndEmail(String username, String emailID) {
```
Comment (AgamAgarwal) [inline src/main/java/dbaccessor/user/UserAccessor.java:37]: Please use `OptionalLong` instead of `Optional<Long>`.
There are specialised classes for optional primitives like `OptionalInt`, `OptionalLong`, etc.

File: src/main/java/controller/UserAPI.java (hunk 18 above)
Comment (NimishaGarg) [reply src/main/java/controller/UserAPI.java]: Done.

File: src/main/java/controller/UserAPI.java (hunk 19 above)
Comment (NimishaGarg) [reply src/main/java/controller/UserAPI.java]: Done.

File: src/main/java/controller/UserAPI.java (hunk 20 above)
Comment (NimishaGarg) [reply src/main/java/controller/UserAPI.java]: Done.

File: src/main/java/googlesignin/GoogleUserMapper.java (hunk 23 above)
Comment (NimishaGarg) [reply src/main/java/googlesignin/GoogleUserMapper.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 24 above)
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/googlesignin/GoogleUser.java (hunk 21 above)
Comment (NimishaGarg) [reply src/main/java/googlesignin/GoogleUser.java:69]: Done.

File: src/main/java/entity/User.java (hunk 22 above)
Comment (NimishaGarg) [reply src/main/java/entity/User.java:79]: Will create a separate PR.👍

File: src/main/java/googlesignin/GoogleUser.java [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+    public Optional<String> getPhoneNumber(String userId) throws IOException {
+        HttpRequestInitializer requestInitializer = new HttpCredentialsAdapter(getCredentials());
+        HttpTransport httpTransport = new NetHttpTransport();
+        JacksonFactory jsonFactory = new JacksonFactory();
+        PeopleService peopleService =
+                new PeopleService.Builder(httpTransport, jsonFactory, requestInitializer).build();
+        Person profile = peopleService.people().get("people/" + userId)
+                .setPersonFields("phoneNumbers")
+                .execute();
+        List<PhoneNumber> phoneNumberList = profile.getPhoneNumbers();
```
Comment (sankalpbansal-google) [inline src/main/java/googlesignin/GoogleUser.java:50]: Try to use Immutable list if possible. Here and elsewhere

File: src/main/java/googlesignin/GoogleUser.java [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+        }
+        String username = GoogleUserMapper.getUsernameFromEmail(email);
+        String picture = null;
+        Object pictureUrl = payload.get("picture");
+        if(pictureUrl != null) {
+            picture = GoogleUserMapper.getPictureFromUrl(pictureUrl.toString());
+        }
+        String userId = payload.getSubject();
+        Optional<String> mobileNo = getPhoneNumber(userId);
+        User newUser;
```
Comment (sankalpbansal-google) [inline src/main/java/googlesignin/GoogleUser.java:80]: It is not a good practice to leave any variable uninitialised. 
Prefer ternary operator.
User newUser= mobileNo.isPresent(): ABC: DEF;

File: src/main/java/googlesignin/GoogleUser.java (hunk 25 above)
Comment (NimishaGarg) [reply src/main/java/googlesignin/GoogleUser.java]: done.

File: src/main/java/googlesignin/GoogleUser.java (hunk 26 above)
Comment (NimishaGarg) [reply src/main/java/googlesignin/GoogleUser.java]: Done.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 27]
```diff
@@ -33,20 +33,18 @@ public long insert(User user) {
...
+    public OptionalLong getUserIdFromEmail(String emailID) {
+        String SQLStatment = "SELECT UserID FROM User WHERE EmailID=@EmailID";
         Statement statement = Statement.newBuilder(SQLStatment)
-                                .bind("Username")
-                                .to(username)
                                 .bind("EmailID")
                                 .to(emailID)
                                 .build();
-        List<User> resultSet = spannerTemplate.query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true));
+        ImmutableList<User> resultSet = ImmutableList.copyOf(spannerTemplate.query(User.class, statement, new SpannerQueryOptions().setAllowPartialRead(true)));
```
Comment (AgamAgarwal) [inline src/main/java/dbaccessor/user/UserAccessor.java:43]: FYI: Since you are never exposing this list outside of this method, you don't really need to make an immutable copy of it.

File: src/main/java/dbaccessor/user/UserAccessor.java [hunk 28]
```diff
@@ -131,26 +129,20 @@ public boolean checkIfUsernameExists(String username) {
...
 
         String SQLStatment = "SELECT * FROM User WHERE UserID=@userId";
         Statement statement = Statement.newBuilder(SQLStatment).bind("userId").to(userId).build();
-        List<User> resultSet = spannerTemplate.query(User.class, statement, null);
- 
-        return resultSet.get(0);
+        return spannerTemplate.query(User.class, statement, null).get(0);
     }
 
     public List<ListChats.UsernameChatId> getUsernameChatIdForSecondUsers(long userId) {
```
Comment (AgamAgarwal) [inline src/main/java/dbaccessor/user/UserAccessor.java:142]: This one however, should return an ImmutableList since the list is being returned.

File: src/main/java/googlesignin/GoogleUserMapper.java [hunk 29]
```diff
@@ -8,18 +12,23 @@
...
     public static String getUsernameFromEmail(String email) {
         int usernameEndsAt = email.indexOf('@');
         if(usernameEndsAt == -1) {
             throw new IllegalArgumentException();
         }
-        return email.substring(0, usernameEndsAt);
+        String baseUsername = email.substring(0, usernameEndsAt);
+        String username = baseUsername;
+        while(userAccessor.checkIfUsernameExists(username)) {
+            username = baseUsername + RandomStringUtils.randomAlphanumeric(3);
```
Comment (AgamAgarwal) [inline src/main/java/googlesignin/GoogleUserMapper.java:29]: Consider leaving a comment here about what this is doing.

File: src/main/java/helper/RequestValidator.java [hunk 30]
```diff
@@ -12,25 +13,25 @@
 
 	private List<String> getMissingFields(Map<String, Object> data, String[] required) {
```
Comment (AgamAgarwal) [inline src/main/java/helper/RequestValidator.java:14]: Ideally this should return ImmutableList instead of the callers of this making a copy.

FYI, you can use ImmutableList.Builder

```
ImmutableList.Builder<String> missing = ImmutableList.builder();

for (String s : required) {
  if (!data.containsKey(s)) {
    missing.add(s);
  }
}

return missing.build();
```

Alternatively, you can also use the streams and collectors.

```
import static com.google.common.collect.ImmutableList.toImmutableList;

return required.stream()
  .filter(s -> !data.containsKey(s))
  .collect(toImmutableList());
```

I think the latter is cleaner.

Comment (AgamAgarwal) [inline src/main/java/helper/RequestValidator.java:14]: The value type here is not relevant or used.
Does this need to be a Map or can it be a Set?

If it has to be a Map, you should use `Map<String, ?>` instead of `Map<String, Object>` to remove restrictions on the callers.

File: src/main/java/helper/RequestValidator.java [hunk 31]
```diff
@@ -12,25 +13,25 @@
...
+                missing.add(s);
             }
         }
         return missing;
     }
 
     public void signupRequestValidator(Map<String, Object> data, String path) {
         String[] required = {"Username", "EmailID", "Password", "MobileNo"};
-        List<String> missing =  getMissingFields(data, required);
+        ImmutableList<String> missing =  ImmutableList.copyOf(getMissingFields(data, required));
```
Comment (AgamAgarwal) [inline src/main/java/helper/RequestValidator.java:26]: As mentioned above, the method `getMissingFields` should really be returning `ImmutableList` instead the caller here making a copy.

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 27 above)
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Changed.

File: src/main/java/dbaccessor/user/UserAccessor.java (hunk 28 above)
Comment (NimishaGarg) [reply src/main/java/dbaccessor/user/UserAccessor.java]: Done.

File: src/main/java/googlesignin/GoogleUserMapper.java (hunk 29 above)
Comment (NimishaGarg) [reply src/main/java/googlesignin/GoogleUserMapper.java:39]: Added

File: src/main/java/helper/RequestValidator.java (hunk 30 above)
Comment (NimishaGarg) [reply src/main/java/helper/RequestValidator.java]: Done.

Comment (NimishaGarg) [reply src/main/java/helper/RequestValidator.java]: Changed to`Set<String>`.

File: src/main/java/helper/RequestValidator.java (hunk 31 above)
Comment (NimishaGarg) [reply src/main/java/helper/RequestValidator.java]: Yes, Done.
